webui/react/src/components/ResponsiveTable.tsx

```diff
@@ -18,8 +19,89 @@ interface Settings {
   tableOffset: number;
 }
 
+interface ContextMenuProps {
+  onVisibleChange: (visible: boolean) => void;
+  record: Record<string, unknown>;
+}
+
+interface ResponsiveTableProps<RecordType> extends TableProps<RecordType> {
+  ContextMenu?: React.FC<ContextMenuProps>;
+  areRowsRightClickable?: boolean;
+  areRowsSelected?: boolean;
+}
+
 /* eslint-disable-next-line @typescript-eslint/ban-types */
-type ResponsiveTable = <T extends object>(props: TableProps<T>) => JSX.Element;
+type ResponsiveTable = <T extends object>(props: ResponsiveTableProps<T>) => JSX.Element;
+
+interface RowProps {
+  ContextMenu: React.FC<ContextMenuProps>;
+  areRowsSelected?: boolean;
+  children?: React.ReactNode;
+  className?: string;
+  record: Record<string, unknown>;
+}
+
+interface CellProps {
+  children?: React.ReactNode;
+  isCellRightClickable?: boolean;
+}
+
+const RightClickableRowContext = createContext({});
+
+const Row = ({
+  className,
+  children,
+  record,
+  ContextMenu,
+  areRowsSelected,
+  ...props
+}: RowProps) => {
+
+  const [ rowHovered, setRowHovered ] = useState(false);
+  const [ rightClickableCellHovered, setRightClickableCellHovered ] = useState(false);
+  const [ contextMenuOpened, setContextMenuOpened ] = useState(false);
+
+  if (areRowsSelected) {
+    return <tr className={className} {...props}>{children}</tr>;
+  }
+
+  const rightClickableCellProps = {
+    onContextMenu: (e : React.MouseEvent) => e.stopPropagation(),
+    onMouseEnter: () => setRightClickableCellHovered(true),
+    onMouseLeave: () => setRightClickableCellHovered(false),
+  };
+
+  const rowContextMenuActive =
```

Comment:
nit: maybe it's just me, but this seems like a misleading variable name. `rowContextMenuActive` = true sounds like the context menu is currently open, not that it's openable.

Reply:
how about `rowContextMenuTriggerableOrOpen` ? Not too unwieldy?

Reply:
That is pretty long, but should be fine.